src/lib.rs

```diff
@@ -1,41 +1,276 @@
+            + (exec::block_timestamp() - self.produced_time) as u128
+                / self.distribution_time as u128
+    }
+
+    /// Updates the reward produced so far and calculates tokens per stake
+    fn update_reward(&mut self) {
+        let reward_produced_at_now = self.produced();
+
+        if reward_produced_at_now > self.reward_produced {
+            let produced_new = reward_produced_at_now - self.reward_produced;
+
+            if self.total_staked > 0 {
+                self.tokens_per_stake = self
+                    .tokens_per_stake
+                    .saturating_add((produced_new * DECIMALS_COUNT) / self.total_staked);
+            }
+
+            self.reward_produced = self.reward_produced.saturating_add(produced_new);
+        }
+    }
+
+    /// Calculates the maximum possible reward
+    /// The reward that the depositor would have received if he had initially paid this amount
+    /// Arguments:
+    /// `amount`: the number of tokens
+    fn get_max_reward(&self, amount: u128) -> u128 {
+        (amount * self.tokens_per_stake) / DECIMALS_COUNT
+    }
+
+    /// Calculates the reward of the staker that is currently avaiable
+    /// The return value cannot be less than zero according to the algorithm
+    fn calc_reward(&mut self) -> u128 {
+        let staker = self
+            .stakers
+            .get(&msg::source())
+            .unwrap_or_else(|| panic!("calc_reward(): Staker {:?} not found", msg::source()));
+
+        self.get_max_reward(staker.balance) + staker.reward_allowed
+            - staker.reward_debt
+            - staker.distributed
+    }
+
+    /// Updates the staking contract.
+    /// Sets the reward to be distributed within distribution time
+    /// param 'config' - updated configuration
+    fn update_staking(&mut self, config: InitStaking) {
+        if config.reward_total == 0 {
+            panic!("update_staking(): reward_total is null");
+        }
+
+        if config.distribution_time == 0 {
+            panic!("update_staking(): distribution_time is null");
+        }
+
+        self.staking_token_address = config.staking_token_address;
+        self.reward_token_address = config.reward_token_address;
+        self.distribution_time = config.distribution_time;
 
-    if payload == "Hello" {
-        msg::reply(b"World", 0).unwrap();
+        self.update_reward();
+        self.all_produced = self.reward_produced;
+        self.produced_time = exec::block_timestamp();
+        self.reward_total = config.reward_total;
+    }
+
+    /// Stakes the tokens
+    /// Arguments:
+    /// `amount`: the number of tokens for the stake
+    async fn stake(&mut self, amount: u128) {
+        if amount == 0 {
+            panic!("stake(): amount is null");
+        }
+
+        let token_address = self.staking_token_address;
+
+        transfer_tokens(&token_address, &msg::source(), &exec::program_id(), amount).await;
+
+        msg::reply(StakingEvent::StakeAccepted(amount), 0).unwrap();
```

Comment:
It is better to move `msg::reply` to the end of the function after all calculations

Reply:
Fixed